File: 0503-Next Greater Element II/c++/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
        
        stack<int> s;
        int length = nums.size();
        vector<int> solution(length, -1);
        
        for(int i = 0; i < 2 * length - 1; ++i) {
            int val = nums[i%length];
            while(!s.empty() && val > nums[s.top()]) {
                solution[s.top()] = val;
                s.pop();
            }
            if(i < length)
                s.push(i);
        }
        return solution;
    }
};

// Runtime: 40 ms, faster than 99.81% of C++ online submissions for Next Greater Element II.
// Memory Usage: 23.8 MB, less than 79.30% of C++ online submissions for Next Greater Element II.

File: 0503-Next Greater Element II/golang/solution.go
package golang

func nextGreaterElements(nums []int) []int {
	var stack []int
	var solution []int
	length := len(nums)

	for idx := 0; idx < length; idx++ {
		solution = append(solution, -1)
	}

	for idx := 0; idx < 2*length-1; idx++ {
		val := nums[idx%length]
		for len(stack) != 0 && val > nums[stack[len(stack)-1]] {
			top := len(stack) - 1
			solution[stack[top]] = val
			stack = stack[:top]
		}

		if idx < length {
			stack = append(stack, idx)
		}
	}

	return solution
}
